File: ports/python/passcode_py/passcode_nodejs_bridge.py
#!/usr/bin/env python3
"""
Python wrapper for Passcode using Node.js bridge
This is a simpler approach that uses the Node.js WASM implementation
"""

import subprocess
import json
from enum import IntEnum
from pathlib import Path


class Algorithm(IntEnum):
    """Supported hash algorithms"""
    SHA3_KMAC_128 = 0
    SHA3_KMAC_256 = 1
    BLAKE3_KEYED_MODE_128 = 2
    BLAKE3_KEYED_MODE_256 = 3


class Passcode:
    """Challenge-response OTP generator"""
    
    def __init__(self, algorithm: Algorithm, key: bytes):
        """
        Create a new Passcode instance
        
        Args:
            algorithm: Algorithm to use (Algorithm enum)
            key: Secret key (bytes, 32 bytes recommended)
        """
        if not isinstance(key, bytes):
            raise TypeError("key must be bytes")
        
        self.algorithm = algorithm
        self.key = key
    
    def compute(self, data: bytes) -> str:
        """
        Compute OTP from challenge data via Node.js
        
        Args:
            data: Challenge data (bytes)
            
        Returns:
            12-character hexadecimal OTP string
        """
        if not isinstance(data, bytes):
            raise TypeError("data must be bytes")
        
        # Call Node.js script
        script = f"""
const {{ Passcode, Algorithm }} = require('{Path(__file__).parent.parent.parent / "nodejs"}');
const key = Buffer.from('{self.key.hex()}', 'hex');
const data = Buffer.from('{data.hex()}', 'hex');
const passcode = new Passcode({int(self.algorithm)}, key);
console.log(passcode.compute(data));
"""
        
        result = subprocess.run(
            ['node', '-e', script],
            capture_output=True,
            text=True,
            check=True
        )
        
        return result.stdout.strip()
    
    def algorithm_name(self) -> str:
        """Get the name of the algorithm"""
        names = {
            Algorithm.SHA3_KMAC_128: "SHA3-KMAC-128",
            Algorithm.SHA3_KMAC_256: "SHA3-KMAC-256",
            Algorithm.BLAKE3_KEYED_MODE_128: "BLAKE3-Keyed-Mode-128",
            Algorithm.BLAKE3_KEYED_MODE_256: "BLAKE3-Keyed-Mode-256",
        }
        return names.get(self.algorithm, "Unknown")

File: ports/dart/lib/src/passcode_nodejs_bridge.dart
import 'dart:io';
import 'dart:typed_data';
import 'algorithm.dart';

/// Challenge-response OTP generator using Node.js WASM bridge
class Passcode {
  /// The algorithm to use
  final Algorithm algorithm;
  
  /// The secret key
  final Uint8List key;
  
  /// Create a new Passcode instance
  /// 
  /// [algorithm] - The hash algorithm to use
  /// [key] - Secret key (32 bytes recommended)
  Passcode(this.algorithm, this.key) {
    if (key.isEmpty) {
      throw ArgumentError('Key cannot be empty');
    }
  }
  
  /// Compute OTP from challenge data
  /// 
  /// [data] - Challenge data
  /// Returns a 12-character hexadecimal OTP string
  Future<String> compute(Uint8List data) async {
    // Use Node.js WASM implementation via subprocess
    final algorithmIndex = algorithm.index;
    final keyHex = _bytesToHex(key);
    final dataHex = _bytesToHex(data);
    
    // Find the Node.js package
    final projectRoot = _findProjectRoot();
    final nodejsPath = '$projectRoot/ports/nodejs';
    
    // Create Node.js script
    final script = '''
const { Passcode, Algorithm } = require('$nodejsPath');
const key = Buffer.from('$keyHex', 'hex');
const data = Buffer.from('$dataHex', 'hex');
const passcode = new Passcode($algorithmIndex, key);
console.log(passcode.compute(data));
''';
    
    // Run Node.js
    final result = await Process.run(
      'node',
      ['-e', script],
    );
    
    if (result.exitCode != 0) {
      throw Exception('Node.js execution failed: ${result.stderr}');
    }
    
    return result.stdout.toString().trim();
  }
  
  /// Get the name of the algorithm
  String get algorithmName => algorithm.name;
  
  String _bytesToHex(Uint8List bytes) {
    return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }
  
  String _findProjectRoot() {
    // Navigate up from ports/dart to find project root
    var current = Directory.current.path;
    
    // Check if we're already in project root
    if (Directory('$current/ports').existsSync()) {
      return current;
    }
    
    // Navigate up
    while (current != '/') {
      if (Directory('$current/ports').existsSync()) {
        return current;
      }
      current = Directory(current).parent.path;
    }
    
    throw Exception('Could not find project root');
  }
}

/// Utility functions

/// Compute BLAKE3-128 hash
Future<String> blake3KeyedMode128(Uint8List key, Uint8List data) async {
  final passcode = Passcode(Algorithm.blake3KeyedMode128, key);
  return await passcode.compute(data);
}

/// Compute BLAKE3-256 hash
Future<String> blake3KeyedMode256(Uint8List key, Uint8List data) async {
  final passcode = Passcode(Algorithm.blake3KeyedMode256, key);
  return await passcode.compute(data);
}

/// Compute SHA3-KMAC-128 hash
Future<String> sha3Kmac128(Uint8List key, Uint8List customization, Uint8List data) async {
  final passcode = Passcode(Algorithm.sha3Kmac128, key);
  return await passcode.compute(data);
}

/// Compute SHA3-KMAC-256 hash
Future<String> sha3Kmac256(Uint8List key, Uint8List customization, Uint8List data) async {
  final passcode = Passcode(Algorithm.sha3Kmac256, key);
  return await passcode.compute(data);
}
